pennylane/__init__.py

```diff
@@ -128,6 +128,8 @@
 from pennylane.interfaces import execute  # pylint:disable=wrong-import-order
 import pennylane.logging  # pylint:disable=wrong-import-order
 
+from pennylane.compiler import qjit, Compiler
```

Comment:
You mention in the description, and in the ADR, that we want a `compiler` module in pennylane, with access to functions such as `qml.compiler.available()`. However, this PR implements these functions in a class, not a module. Not a huge difference, but it is a deviation from the ADR, since now they must be accessed as `qml.Compiler.available()`.

Reply:
I ended up adding these methods to this `Compiler` class just recently because I assumed this static class better addresses 
- the data-hiding principle for the list of available compiler packages (`__BACKENDS`) and the `pennylane.compilers` entry points (`__ENTRY_POINTS`),
- the developer's API abstraction for the JIT compilation support in PL, and  
- avoiding the use of global variables for these properties

What do you think about this?

Reply:
> I ended up adding these methods to this Compiler class just recently because I assumed this static class better addresses:
> - the data-hiding principle for the list of available compiler packages (__BACKENDS) and the pennylane.compilers entry points (__ENTRY_POINTS),

Unfortunately I don't think this makes much of a difference in Python, the attributes are equally accessible in `qml.compiler` as they are in `qml.Compiler`.

> the developer's API abstraction for the JIT compilation support in PL, and

what do you mean by this?

> avoiding the use of global variables for these properties

That's true there are no module globals, instead there are static class variables. As with the first point I'm not sure this makes much of a difference though.

I do see the advantage of a class in terms of code organization, but if the class is the only thing in the module then it's not very relevant.
From a "user" perspective I'd rather use `qml.compiler` than `qml.Compiler`, but the difference is small.
@josh146 do you have any thoughts here?

Reply:
If the main technical motivation is code organization, in that case I would tend to agree with @dime10 here -- as a user/PL developer, I would favour `qml.compiler.available()` over `qml.compiler.Compiler.available()`.

Reply:
@dime10 @josh146 Thanks for the suggestions!

I still think that we can benefit from having a class for the compiler properties, such as names and entry points. What do you think about a simplified `ActiveCompilerData` class to encapsulate the compiler's properties.

``` python
class ActiveCompilerData:
    """This class contains data Compilers data class"""
    
    compilers_dict = dict()

    @classmethod
    def refresh_devices(cls):
        """Scan installed PennyLane compiler pacakges to refresh the compilers names and entry points"""

        reload(pkg_resources)
        for entry in pkg_resources.iter_entry_points("pennylane.compilers"):
            module_name = entry.module_name
            if module_name not in cls.compilers_dict:
                cls.compilers_dict[module_name] = dict()
            cls.compilers_dict[module_name][entry.name] = entry

>>> ActiveCompilerData.refresh_devices()
>>> ActiveCompilerData.compilers_dict
{'catalyst': {'cpl_ext': EntryPoint.parse('cpl_ext = catalyst:pennylane_extensions'),
  'cpl_qjit': EntryPoint.parse('cpl_qjit = catalyst:qjit'),
  'cpl_utils': EntryPoint.parse('cpl_utils = catalyst:utils')}}
``` 

All user facing methods such as `active`, `available`, and other methods will be implemented in the compiler module as the consumers of this class.

Reply:
Sounds good👍